lib/paloaltofw.py

```diff
@@ -0,0 +1,579 @@
+                                     "</member>")
+      address_group_entries.append(self.INDENT * 7 + "</static>")
+      address_group_entries.append(self.INDENT * 7 + "</entry>")
+    address_group_entries.append(self.INDENT * 5 + "</address-group>")
+
+    # SERVICES
+    service = []
+    service.append(self.INDENT * 5 + "<!-- Services -->")
+
+    service.append(self.INDENT * 5 + "<service>")
+    for k, v in Service.service_map.items():
+      service.append(self.INDENT * 6 + '<entry name="' + v["name"] + '">')
+      service.append(self.INDENT * 7 + "<protocol>")
+      service.append(self.INDENT * 8 + "<" + k[1] + ">")
+      tup = str(k[0])[1:-1]
+      if tup[-1] == ",":
+        tup = tup[:-1]
+      service.append(self.INDENT * 9 + "<port>" + tup.replace("'", "") +
+                     "</port>")
+      service.append(self.INDENT * 8 + "</" + k[1] + ">")
+      service.append(self.INDENT * 7 + "</protocol>")
+      service.append(self.INDENT * 6 + "</entry>")
+    service.append(self.INDENT * 5 + "</service>")
+
+    # RULES
+    rules = []
+    rules.append(self.INDENT * 5 + "<!-- Rules -->")
+
+    rules.append(self.INDENT * 5 + "<rulebase>")
+    rules.append(self.INDENT * 6 + "<security>")
+    rules.append(self.INDENT * 7 + "<rules>")
+
+    for name, options in Rule.rules.items():
+      rules.append(self.INDENT * 8 + '<entry name="' + name + '">')
+
+      rules.append(self.INDENT * 9 + "<to>")
+      for tz in options["to_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + tz + "</member>")
+      rules.append(self.INDENT * 9 + "</to>")
+
+      rules.append(self.INDENT * 9 + "<from>")
+      for fz in options["from_zone"]:
+        rules.append(self.INDENT * 10 + "<member>" + fz + "</member>")
+      rules.append(self.INDENT * 9 + "</from>")
+
+      rules.append(self.INDENT * 9 + "<source>")
+      if not options["source"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for s in options["source"]:
+          rules.append(self.INDENT * 10 + "<member>" + s + "</member>")
+      rules.append(self.INDENT * 9 + "</source>")
+
+      rules.append(self.INDENT * 9 + "<destination>")
+      if not options["destination"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for d in options["destination"]:
+          rules.append(self.INDENT * 10 + "<member>" + d + "</member>")
+      rules.append(self.INDENT * 9 + "</destination>")
+
+      rules.append(self.INDENT * 9 + "<service>")
+      if not options["service"]:
+        rules.append(self.INDENT * 10 + "<member>any</member>")
+      else:
+        for s in options["service"]:
+          rules.append(self.INDENT * 10 + "<member>" + s + "</member>")
+      rules.append(self.INDENT * 9 + "</service>")
+
+      rules.append(self.INDENT * 9 + "<action>" + Term.ACTIONS.get(
+          str(options["action"])) + "</action>")
+
+      if fz == tz == "any":
```

Comment:
@ankenyr
Not sure why it's originally set up that way, but it can be related to the fact that PAN allows rules to have several source/destination zones.
Right now line 149 
```
self.options["from_zone"] = [from_zone]
```
sets it to an array containing only 1 element.
But we could for example imagine it (in the future) being a list from the .pol file header, line 292
```
self.from_zone = filter_options[1].split(',')
```
(or something similar).

About the possibility of fz or tz being empty, I believe we can add a check after line 150:
```
    if not from_zone or not to_zone:
        raise PaloAltoFWOptionError("Source or destination zone is empty.")
```